File: apps/android/app/src/main/kotlin/com/wordocious/app/data/VSMatchService.kt
package com.wordocious.app.data

import android.os.Handler
import android.os.Looper
import io.socket.client.IO
import io.socket.client.Socket
import kotlinx.serialization.json.Json
import org.json.JSONObject

/**
 * Thin socket.io wrapper for VS — the Android equivalent of iOS VSMatchService /
 * web SocketIOMatchService. Owns the connection and exposes typed emit methods +
 * inbound-event callbacks. The VS state machine lives in VSMatchViewModel. All
 * inbound callbacks are marshalled to the main thread so the UI can update directly.
 */
class VSMatchService : VSTransport {
    private var socket: Socket? = null
    private val main = Handler(Looper.getMainLooper())
    private val json = Json { ignoreUnknownKeys = true; coerceInputValues = true }

    // Inbound callbacks (set by the view model)
    override var onConnect: (() -> Unit)? = null
    override var onDisconnect: (() -> Unit)? = null
    override var onQueueStatus: ((VSQueueStatus) -> Unit)? = null
    override var onMatchFound: ((VSMatchFound) -> Unit)? = null
    override var onMatchStart: ((VSMatchStart) -> Unit)? = null
    override var onGuessResult: ((VSGuessResult) -> Unit)? = null
    override var onOpponentProgress: ((VSOpponentProgress) -> Unit)? = null
    override var onMatchEnded: ((VSMatchEnded) -> Unit)? = null
    override var onOpponentStageCompleted: ((VSStageEvent) -> Unit)? = null
    override var onRematchOffered: (() -> Unit)? = null
    override var onRematchDeclined: (() -> Unit)? = null
    override var onRematchStart: ((VSRematchStart) -> Unit)? = null
    override var onOpponentLeft: (() -> Unit)? = null
    override var onServerError: ((VSServerError) -> Unit)? = null
    override var onOpponentTyping: (() -> Unit)? = null

    override val isConfigured: Boolean get() = VSConfig.isConfigured

    // ── Connection ──────────────────────────────────────────────────────────────
    override fun connect(presenceId: String?) {
        val opts = IO.Options().apply {
            forceNew = true
            reconnection = true
            transports = arrayOf("websocket")
            // socket.io v3+ handshake auth — server reads handshake.auth.presenceId.
            if (presenceId != null) auth = mapOf("presenceId" to presenceId)
        }
        val s = runCatching { IO.socket(VSConfig.SERVER_URL, opts) }.getOrNull() ?: return
        socket = s
        register(s)
        s.connect()
    }

    override fun disconnect() {
        socket?.let { it.off(); it.disconnect() }
        socket = null
    }

    // ── Outbound (client → server) ────────────────────────────────────────────────
    override fun joinQueue(mode: String, dailySeed: String?, inviteCode: String?) {
        val payload = JSONObject().put("mode", mode)
        dailySeed?.let { payload.put("dailySeed", it) }
        inviteCode?.let { payload.put("inviteCode", it) }
        socket?.emit(VSEvent.JOIN_QUEUE, payload)
    }

    override fun leaveQueue() { socket?.emit(VSEvent.LEAVE_QUEUE) }

    override fun submitGuess(guess: String, boardIndex: Int) {
        socket?.emit(VSEvent.SUBMIT_GUESS, JSONObject().put("guess", guess).put("boardIndex", boardIndex))
    }

    override fun boardSolved(boardIndex: Int) {
        socket?.emit(VSEvent.BOARD_SOLVED, JSONObject().put("boardIndex", boardIndex))
    }

    override fun playerCompleted(status: String, totalGuesses: Int, timeMs: Int) {
        socket?.emit(VSEvent.PLAYER_COMPLETED, JSONObject().put("status", status).put("totalGuesses", totalGuesses).put("timeMs", timeMs))
    }

    override fun stageCompleted(stageIndex: Int) {
        socket?.emit(VSEvent.STAGE_COMPLETED, JSONObject().put("stageIndex", stageIndex))
    }

    /** Throttle upstream (VSMatchViewModel) — at most one ping per 1.5s. */
    override fun emitTyping() { socket?.emit(VSEvent.TYPING) }

    override fun abandonMatch() { socket?.emit(VSEvent.ABANDON_MATCH) }
    override fun offerRematch() { socket?.emit(VSEvent.OFFER_REMATCH) }
    override fun declineRematch() { socket?.emit(VSEvent.DECLINE_REMATCH) }

    // ── Handler registration ────────────────────────────────────────────────────
    private fun register(s: Socket) {
        s.on(Socket.EVENT_CONNECT) { runOnMain { onConnect?.invoke() } }
        s.on(Socket.EVENT_DISCONNECT) { runOnMain { onDisconnect?.invoke() } }

        bind(s, VSEvent.QUEUE_STATUS, VSQueueStatus.serializer()) { onQueueStatus?.invoke(it) }
        bind(s, VSEvent.MATCH_FOUND, VSMatchFound.serializer()) { onMatchFound?.invoke(it) }
        bind(s, VSEvent.MATCH_START, VSMatchStart.serializer()) { onMatchStart?.invoke(it) }
        bind(s, VSEvent.GUESS_RESULT, VSGuessResult.serializer()) { onGuessResult?.invoke(it) }
        bind(s, VSEvent.OPPONENT_PROGRESS, VSOpponentProgress.serializer()) { onOpponentProgress?.invoke(it) }
        bind(s, VSEvent.MATCH_ENDED, VSMatchEnded.serializer()) { onMatchEnded?.invoke(it) }
        bind(s, VSEvent.OPPONENT_STAGE_COMPLETED, VSStageEvent.serializer()) { onOpponentStageCompleted?.invoke(it) }
        bind(s, VSEvent.REMATCH_START, VSRematchStart.serializer()) { onRematchStart?.invoke(it) }
        bind(s, VSEvent.ERROR, VSServerError.serializer()) { onServerError?.invoke(it) }

        s.on(VSEvent.REMATCH_OFFERED) { runOnMain { onRematchOffered?.invoke() } }
        s.on(VSEvent.REMATCH_DECLINED) { runOnMain { onRematchDeclined?.invoke() } }
        s.on(VSEvent.OPPONENT_LEFT) { runOnMain { onOpponentLeft?.invoke() } }
        s.on(VSEvent.OPPONENT_TYPING) { runOnMain { onOpponentTyping?.invoke() } }
    }

    /** Decode args[0] (a JSONObject) into [T] and deliver on the main thread. */
    private fun <T> bind(s: Socket, event: String, deserializer: kotlinx.serialization.DeserializationStrategy<T>, deliver: (T) -> Unit) {
        s.on(event) { args ->
            val obj = args.firstOrNull() as? JSONObject ?: return@on
            val value = runCatching { json.decodeFromString(deserializer, obj.toString()) }.getOrNull() ?: return@on
            runOnMain { deliver(value) }
        }
    }

    private fun runOnMain(block: () -> Unit) {
        if (Looper.myLooper() == Looper.getMainLooper()) block() else main.post(block)
    }
}

File: apps/ios/Wordocious/Sources/VSMatchService.swift
import Foundation
import SocketIO

/// Thin socket.io wrapper for VS — the native equivalent of
/// apps/web/lib/adapters/match-service.ts (SocketIOMatchService). It owns the
/// connection and exposes typed emit methods + inbound-event closures. The VS
/// state machine lives in VSMatchViewModel. All inbound closures fire on the
/// main actor so the UI can update directly.
final class VSMatchService {
    private var manager: SocketManager?
    private var socket: SocketIOClient?

    // MARK: Inbound callbacks (set by the view model)
    var onConnect: (() -> Void)?
    var onDisconnect: (() -> Void)?
    var onQueueStatus: ((VSQueueStatus) -> Void)?
    var onMatchFound: ((VSMatchFound) -> Void)?
    var onMatchStart: ((VSMatchStart) -> Void)?
    var onGuessResult: ((VSGuessResult) -> Void)?
    var onOpponentProgress: ((VSOpponentProgress) -> Void)?
    var onMatchEnded: ((VSMatchEnded) -> Void)?
    var onOpponentStageCompleted: ((VSStageEvent) -> Void)?
    var onRematchOffered: (() -> Void)?
    var onRematchDeclined: (() -> Void)?
    var onRematchStart: ((VSRematchStart) -> Void)?
    var onOpponentLeft: (() -> Void)?
    var onOpponentTyping: (() -> Void)?
    var onServerError: ((VSServerError) -> Void)?

    var isConfigured: Bool { VSConfig.isConfigured }

    // MARK: Connection

    func connect(presenceId: String?) {
        guard let url = VSConfig.serverURL else { return }
        let manager = SocketManager(socketURL: url, config: [
            .log(false), .compress, .forceWebsockets(true), .reconnects(true),
        ])
        self.manager = manager
        let socket = manager.defaultSocket
        self.socket = socket
        register(socket)
        socket.connect(withPayload: presenceId.map { ["presenceId": $0] })
    }

    func disconnect() {
        socket?.disconnect()
        manager?.disconnect()
        socket = nil
        manager = nil
    }

    // MARK: Outbound (client → server)

    func joinQueue(mode: String, dailySeed: String?, inviteCode: String?) {
        var payload: [String: Any] = ["mode": mode]
        if let dailySeed { payload["dailySeed"] = dailySeed }
        if let inviteCode { payload["inviteCode"] = inviteCode }
        socket?.emit(VSEvent.joinQueue, payload)
    }

    func leaveQueue() { socket?.emit(VSEvent.leaveQueue) }

    func submitGuess(_ guess: String, boardIndex: Int = 0) {
        socket?.emit(VSEvent.submitGuess, ["guess": guess, "boardIndex": boardIndex])
    }

    func boardSolved(boardIndex: Int) { socket?.emit(VSEvent.boardSolved, ["boardIndex": boardIndex]) }

    func playerCompleted(status: String, totalGuesses: Int, timeMs: Int) {
        socket?.emit(VSEvent.playerCompleted, ["status": status, "totalGuesses": totalGuesses, "timeMs": timeMs])
    }

    func stageCompleted(stageIndex: Int) { socket?.emit(VSEvent.stageCompleted, ["stageIndex": stageIndex]) }
    /// Throttled activity ping (the caller throttles) — relayed as opponent_typing.
    func emitTyping() { socket?.emit(VSEvent.typing) }
    func abandonMatch() { socket?.emit(VSEvent.abandonMatch) }
    func offerRematch() { socket?.emit(VSEvent.offerRematch) }
    func declineRematch() { socket?.emit(VSEvent.declineRematch) }

    // MARK: Handler registration

    private func register(_ socket: SocketIOClient) {
        socket.on(clientEvent: .connect) { [weak self] _, _ in self?.main { self?.onConnect?() } }
        socket.on(clientEvent: .disconnect) { [weak self] _, _ in self?.main { self?.onDisconnect?() } }

        bind(socket, VSEvent.queueStatus, VSQueueStatus.self) { [weak self] in self?.onQueueStatus?($0) }
        bind(socket, VSEvent.matchFound, VSMatchFound.self) { [weak self] in self?.onMatchFound?($0) }
        bind(socket, VSEvent.matchStart, VSMatchStart.self) { [weak self] in self?.onMatchStart?($0) }
        bind(socket, VSEvent.guessResult, VSGuessResult.self) { [weak self] in self?.onGuessResult?($0) }
        bind(socket, VSEvent.opponentProgress, VSOpponentProgress.self) { [weak self] in self?.onOpponentProgress?($0) }
        bind(socket, VSEvent.matchEnded, VSMatchEnded.self) { [weak self] in self?.onMatchEnded?($0) }
        bind(socket, VSEvent.opponentStageCompleted, VSStageEvent.self) { [weak self] in self?.onOpponentStageCompleted?($0) }
        bind(socket, VSEvent.rematchStart, VSRematchStart.self) { [weak self] in self?.onRematchStart?($0) }
        bind(socket, VSEvent.error, VSServerError.self) { [weak self] in self?.onServerError?($0) }

        // Payload-less events.
        socket.on(VSEvent.rematchOffered) { [weak self] _, _ in self?.main { self?.onRematchOffered?() } }
        socket.on(VSEvent.rematchDeclined) { [weak self] _, _ in self?.main { self?.onRematchDeclined?() } }
        socket.on(VSEvent.opponentLeft) { [weak self] _, _ in self?.main { self?.onOpponentLeft?() } }
        socket.on(VSEvent.opponentTyping) { [weak self] _, _ in self?.main { self?.onOpponentTyping?() } }
    }

    /// Decode the first element of a socket payload into `T` and deliver on main.
    private func bind<T: Decodable>(_ socket: SocketIOClient, _ event: String, _ type: T.Type,
                                    _ deliver: @escaping (T) -> Void) {
        socket.on(event) { [weak self] data, _ in
            guard let value: T = Self.decode(type, data) else { return }
            self?.main { deliver(value) }
        }
    }

    /// Shared decoder for inbound socket events — JSONDecoder is safe to reuse
    /// for sequential decodes on one thread (socket.io delivers handlers on its
    /// single handleQueue), and reusing it avoids a fresh allocation per event
    /// on the hot in-match path (guess_result / opponent_progress).
    private static let eventDecoder = JSONDecoder()

    private static func decode<T: Decodable>(_ type: T.Type, _ data: [Any]) -> T? {
        // JSONSerialization.data(withJSONObject:) RAISES an NSException (not a
        // Swift error, so `try?` can't catch it → crash) when `first` isn't a
        // valid top-level JSON object — e.g. a String/Number/NSNull error
        // payload. Guard with isValidJSONObject so a malformed/unexpected
        // socket payload is skipped instead of aborting the app.
        guard let first = data.first,
              JSONSerialization.isValidJSONObject(first),
              let json = try? JSONSerialization.data(withJSONObject: first) else { return nil }
        return try? eventDecoder.decode(T.self, from: json)
    }

    private func main(_ work: @escaping () -> Void) {
        if Thread.isMainThread { work() } else { DispatchQueue.main.async(execute: work) }
    }
}
